drivers/usb/host/xhci-mem.c

```diff
@@ -1477,9 +1480,35 @@ int xhci_endpoint_init(struct xhci_hcd *xhci,
 
 	mult = xhci_get_endpoint_mult(udev, ep);
 	max_packet = usb_endpoint_maxp(&ep->desc);
-	max_burst = xhci_get_endpoint_max_burst(udev, ep);
 	avg_trb_len = max_esit_payload;
 
+	/*
+	 * VL805 errata - Bulk OUT bursts to superspeed mass-storage
+	 * devices behind hub ports can cause data corruption with
+	 * non-wMaxPacket-multiple transfers.
+	 */
+	for (i = 0; i < udev->config->desc.bNumInterfaces; i++) {
+		intfc = udev->config->intf_cache[i];
+		/*
+		 * Slight hack - look at interface altsetting 0, which
+		 * should be the UMS bulk-only interface. If the class
+		 * matches, then we disable out bursts for all OUT
+		 * endpoints because endpoint assignments may change
+		 * between alternate settings.
+		 */
+		if (intfc->altsetting[0].desc.bInterfaceClass ==
+		    USB_CLASS_MASS_STORAGE) {
```

Comment:
It doesn't make sense this only applies to mass storage!

Reply:
Sorta. Minimum tweak, though. The actual fix needs to be in the VL805 firmware, I guess. 
Another "fun" fact: I am following and testing this out of curiosity. On an RPi 4B 1.5 with 8GB RAM this works fine (as it seems, with an Argon 40 m.2 SATA enclosure and 174c:1156 ASMedia adapter). On a CM4 I/O board with WaveShare USB3.2 extension card (also VL805, but apparrently other firmware), I don't get the corruption, but uas sporadic errors with device resets. Need to disable uas for the external drives (JMicron and UGreen adapters). This also slows things down a bit.

Reply:
By chance, could you run the same tests under FreeBSD? We have a different USB stack and XHCI host controller!
We have some pre-built images here, which you can easily boot the RPI4 from:
https://download.freebsd.org/ftp/snapshots/arm64/aarch64/ISO-IMAGES/13.1/

Reply:
> It doesn't make sense this only applies to mass storage!

Do you have an example of a superspeed device that is not a mass-storage class device which is a victim of this bug? Two factors are required - a driver that enqueues a vulnerable sequence of TRBs, and a device that both supports bursts and frequently signals flow control on its OUT endpoints.

Reply:
Well, I did - still searching for my micro HDMI to HDMI cable... anyway. I didn't come that far, just trying to mount the ntfs volume fails with command errors. wich is sorta consistent with VL805 firmware.

![debian_uas_errors](https://user-images.githubusercontent.com/33185912/190925876-e392558f-20b4-4d1a-ab61-f7688824e349.jpg)
![freebsd_trymount](https://user-images.githubusercontent.com/33185912/190925879-1fd00c5f-5a33-4686-a1b5-924cf31491ab.jpg)

That the JMicron adapter throws errors with uas is well known. Couldn't mount the BTRFS UGreen device on FreeBSD. Anyway, the UGreen device is mounted and used as uas without errors on another RPi 4B 1.5 8GB.

![grafik](https://user-images.githubusercontent.com/33185912/190926164-97030459-1ccf-4430-99d9-5f2015931095.png)

since I have a FreeBSD SD card and a setup with keyboard/screen now, I can do other tests. It looks to me like this actually is a firmware issue with the VL805.

Reply:
Isn't this workaround related also to the wireless cards? I've noticed that a mt76 driver maintainer has suggested usage of @P33M's workaround because of Mediatek MT7612U random failures ([link](https://github.com/openwrt/mt76/issues/405#issuecomment-1066549535)). AFAIR those issues were mostly found on Raspberry Pi 4.
I've noticed similar issues on my setup (RPi4 + MT7612U). Disabling scatter-gatter (_disable_usb_sg_ parameter in _mt76_usb_ module) on mt76 seems to drop occurrence of those failures but in my case it didn't solve it ultimately (crashes occurred like once a week).

Reply:
This would be the one case you need to disable bursts for all burst-capable devices behind a hub. Bad news.

Reply:
Is there a descriptor which tells how many bursts a given USB 3.0 HUB supports? I'm trying to look in the USB specification.

Reply:
@dobo90 if you think that the failure is related, then please create a new issue. Try building a custom kernel from this repo with the above interface class check disabled, and see if your wireless dongle still crashes.

https://www.raspberrypi.com/documentation/computers/linux_kernel.html#building-the-kernel

Reply:
Hubs don't need to know protocol layer details about the packets they are forwarding. They are store-and-forward, and manage link state and packet credit between upstream and downstream ports. A data burst simply refers to the number of outstanding DPs that a transfer source can post before it has to receive an ACK from the destination, and a "burst" can continue indefinitely if there is no other pending endpoint activity. The rules for starting and ending a burst are only functions of the controller and the device.

Reply:
It's called bMaxBurst, but it's not defined for HUBs (yet).

Reply:
@dobo90 does your MT7612U support bursting? If not, how could this fix be related? I have a MT7921AU (CF-953AX) and verified via `sudo lsusb -vv | grep -E 'bInterfaceClass|Burst'` that `bMaxBurst` is 0. If I understand the current fix correctly, it simply forces `max_burst = 0` if a USB_CLASS_MASS_STORAGE device is present on a hub (which should include the root_hub, and thus all usb devices plugged into the RPi4?). But if `bMaxBurst` is already 0, that should not have any effect.

And in fact, I'm able to reproduce firmware crashes on CF-953AX on 6.0.0-rc7-v8+ which includes the current fix, both in the case when a USB_CLASS_MASS_STORAGE device is present during boot and remains present, as well as if no USB_CLASS_MASS_STORAGE device is present during boot or at any time.

Reply:
Having a discussion on a merge request line comment isn't how github is supposed to be used - if you have a bug that is related to this commit, then please create a new issue, fill out the template, and refer to this with a link.